nengo_dl/converter.py

```diff
@@ -0,0 +1,1397 @@
+
+        # perform custom checks in layer converters
+        if ConverterClass is None:
+            error_msg = None
+        else:
+            convertible, error_msg = ConverterClass.convertible(layer, self)
+            if not convertible:
+                ConverterClass = None
+
+        if ConverterClass is None:
+            # this means that there is no LayerConverter compatible with this layer
+            # (either because it has an unknown type, or it failed the ``.convertible``
+            # check due to its internal parameterization)
+            if self.allow_fallback:
+                warnings.warn(
+                    "%sFalling back to TensorNode."
+                    % (error_msg + ". " if error_msg else "")
+                )
+                ConverterClass = self.converters[None]
+            else:
+                raise TypeError(
+                    "%sUnable to convert layer %s to native Nengo objects; set "
+                    "allow_fallback=True if you would like to use a TensorNode "
+                    "instead, or consider registering a custom LayerConverter for this "
+                    "layer type." % (error_msg + ". " if error_msg else "", layer.name)
+                )
+
+        converter = ConverterClass(layer, self)
+
+        self._layer_converters[layer] = converter
+
+        return converter
+
+    @classmethod
+    def register(cls, keras_layer):
+        """
+        A decorator for adding a class to the converter registry.
+
+        Parameters
+        ----------
+        keras_layer : ``tf.keras.layers.Layer``
+            The Layer associated with the conversion function being registered.
+        """
+
+        def register_converter(convert_cls):
+            if keras_layer in cls.converters:
+                warnings.warn(
+                    "Layer '%s' already has a converter. Overwriting." % keras_layer
+                )
+
+            cls.converters[keras_layer] = convert_cls
+
+            return convert_cls
+
+        return register_converter
+
+    class TensorDict:
+        """A dictionary-like object that works with TensorFlow Tensors."""
+
+        def __init__(self):
+            self.dict = collections.OrderedDict()
+
+        def __setitem__(self, key, val):
+            if isinstance(key, tf.Tensor):
+                key = key.experimental_ref()
+
+            self.dict[key] = val
+
+        def __getitem__(self, key):
+            if isinstance(key, tf.Tensor):
+                key = key.experimental_ref()
+
+            return self.dict[key]
+
+
+class LayerConverter:
+    """
+    Base class for converter classes, which contain the logic for mapping some Keras
+    layer type to Nengo objects.
+
+    Parameters
+    ----------
+    layer : ``tf.keras.layers.Layer``
+        The Layer object being converted.
+    converter : `.Converter`
+        The parent Converter class running the conversion process.
+    """
```

Comment:
Are any of the class parameters worth documenting here? I noticed `has_weights` below and had to search around to find it. I know it's likely mostly a dev-facing thing, but you do suggest that users implement subclasses of this, so it might be worthwhile.

There isn't much support in `numpydoc` for class parameters, so I would maybe do it mainly in the main text before the `Parameters` section, if any of them are worth advertising to users.